packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -224,13 +231,48 @@ class TabController extends ChangeNotifier {
           notifyListeners();
         });
     } else {
+      final bool wasIndexChanging = _indexIsChangingCount > 0;
       _indexIsChangingCount += 1;
+      // should only notify AnimationStartedListener when
+      // it was not called by ScrollEndNotification or
+      // index was changing before (so that TabBar and TabBarView will start a new animation with new target index.)
+      // This happens when controller.index = ? is called while animation was ongoing.
+      if (_animationController.value != _index.toDouble() || wasIndexChanging) {
+        _notifyAnimationStartedListeners(kTabScrollDuration, Curves.ease);
+      }
       _animationController.value = _index.toDouble();
       _indexIsChangingCount -= 1;
       notifyListeners();
     }
   }
 
+  void _notifyAnimationStartedListeners(Duration duration, Curve curve) {
+    final List<AnimationStartedListener> localListeners = List<AnimationStartedListener>.from(_animationStartedListeners);
+    for (final AnimationStartedListener listener in localListeners) {
+      if (_animationStartedListeners.contains(listener))
+        listener(duration, curve);
+    }
+  }
+
+  /// Calls listener every time the animation starts by setting [index] or [animateTo].
+  /// When setting [index], duration is always [kTabScrollDuration] and
```

Comment:
is this true? if you set the index directly, the duration should be null which should cause a warping effect on the tabbarview

Reply:
we should also mention the duration maybe null, and what does it mean to be null

Reply:
Oh, the way I implemented it was to pass in ```kTabScrollDuration``` and ```Curves.ease``` to AnimationStartedListener when we're just setting the index:
![image](https://user-images.githubusercontent.com/4879766/89485627-2adb1800-d7dc-11ea-9db9-2fb611b94bbe.png)

So the duration and curve given to AnimationStartedListener are effectively NonNull. Would it be better we pass in null instead?

Reply:
I think we need to put null to correctly reflect what is actually happening inside tabcontroller. it is up to the observers to decide how to react to null duration(for example giving it kTabScrollDuration)